### ISSUE TITLE
nodetool removenode hangs in repair, waiting on streaming semaphore

### ISSUE THREAD

**Issue by kostja** [scylladb/scylladb#17591]
*Installation details*
Scylla version (or git commit hash): current master
Cluster size: 5 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): Fedora 38

I am using https://github.com/kostja/scylla-ansible-demo to set up and start a cluster of 5 nodes on my local machine, IP range 127.0.0.[1-5]

I bring 3 out of 5 nodes down.

I follow https://opensource.docs.scylladb.com/stable/architecture/raft.html#recover-raft-procedure to restore the cluster from the majority loss.

Specifially:
- I update the raft group0 state key to 'recovery' on the two remaining nodes
- I perform a rolling restart of the nodes.

The next step of the recovery procedure is to remove the remaining dead nodes.

I execute nodetool removenode (I use scylla-nodetool, to avoid scylla-jmx, however, the bug reproduces with Java nodetool as well) --ignore-dead-nodes, and on first, second, or third node removal it **hangs forever.**

Impact: this renders the majority-loss recovery procedure fully unusable.

**Comment by kostja** [comment]
Last messages from repair:
```
DEBUG 2024-02-29 22:25:02,388 [shard 1:strm] repair - repair[8f70d7dc-1686-4fc9-943d-c00ef9f0390f]: node ops progress bootstrap=1, replace=1, rebuild=1, decom
mission=1, removenode=0.037694015, repair=1
DEBUG 2024-02-29 22:25:02,389 [shard 1:strm] repair - <<< Finished Row Level Repair (Master): local=127.0.0.1, peers={127.0.0.2}, repair_meta_id=63, keyspace=
system_distributed_everywhere, cf=cdc_generation_descriptions_v2, range=(-404417379823617528,165113280337227894], tx_hashes_nr=0, rx_hashes_nr=0, tx_row_nr=0,
 rx_row_nr=0, row_from_disk_bytes={{127.0.0.1, 0}, {127.0.0.2, 0}}, row_from_disk_nr={{127.0.0.1, 0}, {127.0.0.2, 0}}
DEBUG 2024-02-29 22:25:02,389 [shard 1:strm] repair - <<< Finished Row Level Repair (Master): local=127.0.0.1, peers={127.0.0.2}, repair_meta_id=64, keyspace=
system_distributed_everywhere, cf=cdc_generation_descriptions_v2, range=(165113280337227894,219829027767488212], tx_hashes_nr=0, rx_hashes_nr=0, tx_row_nr=0, 
rx_row_nr=0, row_from_disk_bytes={{127.0.0.1, 0}, {127.0.0.2, 0}}, row_from_disk_nr={{127.0.0.1, 0}, {127.0.0.2, 0}}
DEBUG 2024-02-29 22:25:02,389 [shard 1:strm] repair - <<< Finished Row Level Repair (Master): local=127.0.0.1, peers={127.0.0.2}, repair_meta_id=65, keyspace=
system_distributed_everywhere, cf=cdc_generation_descriptions_v2, range=(219829027767488212,531618459526350624], tx_hashes_nr=0, rx_hashes_nr=0, tx_row_nr=0, 
rx_row_nr=0, row_from_disk_bytes={{127.0.0.1, 0}, {127.0.0.2, 0}}, row_from_disk_nr={{127.0.0.1, 0}, {127.0.0.2, 0}}
DEBUG 2024-02-29 22:25:02,389 [shard 1:strm] repair - repair[8f70d7dc-1686-4fc9-943d-c00ef9f0390f]: node ops progress bootstrap=1, replace=1, rebuild=1, decom
mission=1, removenode=0.039911307, repair=1
DEBUG 2024-02-29 22:25:02,389 [shard 1:strm] repair - repair[8f70d7dc-1686-4fc9-943d-c00ef9f0390f]: node ops progress bootstrap=1, replace=1, rebuild=1, decom
mission=1, removenode=0.042128604, repair=1
DEBUG 2024-02-29 22:25:02,389 [shard 1:strm] repair - repair[8f70d7dc-1686-4fc9-943d-c00ef9f0390f]: node ops progress bootstrap=1, replace=1, rebuild=1, decom
mission=1, removenode=0.044345897, repair=1
```

**Comment by kbr-scylla** [comment]
I couldn't reproduce it on my laptop or Warsaw machines.
But I could reproduce it on your hulk machine.

You don't need to go through recovery mode or start cluster in `consistent-topology-changes` experimental mode, the issue is reproducible as follows: start 5 node cluster, shutdown one node, try to removenode it. Very simple. Maybe works with fewer nodes too, haven't checked.

I determined that repair is getting stuck on followers in `repair_service::insert_repair_meta` call, here:
```cpp
        return db.local().obtain_reader_permit(cf, "repair-meta", db::no_timeout, {}).then([s = std::move(s),
```

easy to see if you change `db::no_timeout` to something like `db::timeout_clock::now() + std::chrono::minutes{1}`.

We recently saw repairs getting stuck on customer clusters, which caused other operations that used the streaming scheduling group to eventually time out: https://github.com/scylladb/scylla-enterprise/issues/3876

Could be the same root cause.

Posting semaphore diagnostic dumps (there were 5 nodes, 127.0.0.2 was being removed):
127.0.0.1:
```
INFO  2024-03-04 18:19:33,431 [shard 0:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 163840/1252838604 memory resources: timed out, dumping permit diagnostics:
permits count   memory  table/operation/state
10      10      160K    system_distributed.cdc_streams_descriptions_v2/repair-meta/active
2       0       0B      system_distributed.service_levels/repair-meta/waiting_for_admission
10      0       0B      system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/waiting_for_admission
7       0       0B      system_distributed.cdc_streams_descriptions_v2/repair-meta/waiting_for_admission

29      10      160K    total

Stats:
permit_based_evictions: 0
time_based_evictions: 0
inactive_reads: 0
total_successful_reads: 646
total_failed_reads: 0
total_reads_shed_due_to_overload: 0
total_reads_killed_due_to_kill_limit: 0
reads_admitted: 1919
reads_enqueued_for_admission: 439
reads_enqueued_for_memory: 0
reads_admitted_immediately: 1499
reads_queued_because_ready_list: 392
reads_queued_because_need_cpu_permits: 0
reads_queued_because_memory_resources: 47
reads_queued_because_count_resources: 0
reads_queued_with_eviction: 0
total_permits: 1938
current_permits: 29
need_cpu_permits: 0
awaits_permits: 0
disk_reads: 0
sstables_read: 0
```
+
```
INFO  2024-03-04 18:19:33,439 [shard 1:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 169108/1252838604 memory resources: timed out, dumping permit diagnostics:
permits count   memory  table/operation/state
12      10      165K    system_distributed.cdc_streams_descriptions_v2/repair-meta/active
10      0       0B      system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/waiting_for_admission
2       0       0B      system_distributed.cdc_streams_descriptions_v2/repair-meta/waiting_for_admission

24      10      165K    total

Stats:
permit_based_evictions: 0
time_based_evictions: 0
inactive_reads: 0
total_successful_reads: 1
total_failed_reads: 0
total_reads_shed_due_to_overload: 0
total_reads_killed_due_to_kill_limit: 0
reads_admitted: 1282
reads_enqueued_for_admission: 43
reads_enqueued_for_memory: 0
reads_admitted_immediately: 1251
reads_queued_because_ready_list: 0
reads_queued_because_need_cpu_permits: 0
reads_queued_because_memory_resources: 43
reads_queued_because_count_resources: 0
reads_queued_with_eviction: 0
total_permits: 1294
current_permits: 24
need_cpu_permits: 0
awaits_permits: 0
disk_reads: 0
sstables_read: 0
```

127.0.0.3:
```
INFO  2024-03-04 18:19:33,422 [shard 1:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 163840/1252838604 memory resources: timed out, dumping permit diagnostics:
permits count   memory  table/operation/state
10      10      160K    system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/active
3       0       0B      system_distributed.view_build_status/repair-meta/waiting_for_admission
1       0       0B      system_traces.node_slow_log/repair-meta/waiting_for_admission
36      0       0B      system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/waiting_for_admission

50      10      160K    total

Stats:
permit_based_evictions: 0
time_based_evictions: 0
inactive_reads: 0
total_successful_reads: 2
total_failed_reads: 0
total_reads_shed_due_to_overload: 0
total_reads_killed_due_to_kill_limit: 0
reads_admitted: 1129
reads_enqueued_for_admission: 316
reads_enqueued_for_memory: 0
reads_admitted_immediately: 853
reads_queued_because_ready_list: 0
reads_queued_because_need_cpu_permits: 0
reads_queued_because_memory_resources: 316
reads_queued_because_count_resources: 0
reads_queued_with_eviction: 0
total_permits: 1169
current_permits: 50
need_cpu_permits: 0
awaits_permits: 0
disk_reads: 0
sstables_read: 0
```
+
```
INFO  2024-03-04 18:19:33,433 [shard 0:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 163840/1252838604 memory resources: timed out, dumping permit diagnostics:
permits count   memory  table/operation/state
10      10      160K    system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/active
6       0       0B      system_distributed.service_levels/repair-meta/waiting_for_admission
36      0       0B      system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/waiting_for_admission
5       0       0B      system_distributed.cdc_streams_descriptions_v2/repair-meta/waiting_for_admission

57      10      160K    total

Stats:
permit_based_evictions: 0
time_based_evictions: 0
inactive_reads: 0
total_successful_reads: 1189
total_failed_reads: 0
total_reads_shed_due_to_overload: 0
total_reads_killed_due_to_kill_limit: 0
reads_admitted: 2318
reads_enqueued_for_admission: 1250
reads_enqueued_for_memory: 0
reads_admitted_immediately: 1115
reads_queued_because_ready_list: 922
reads_queued_because_need_cpu_permits: 5
reads_queued_because_memory_resources: 323
reads_queued_because_count_resources: 0
reads_queued_with_eviction: 0
total_permits: 2365
current_permits: 57
need_cpu_permits: 0
awaits_permits: 0
disk_reads: 0
sstables_read: 0
```

127.0.0.4:
```
INFO  2024-03-04 18:19:33,434 [shard 0:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 163840/1252838604 memory resources: timed out, dumping permit diagnostics:
permits count   memory  table/operation/state
6       6       96K     system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/active
4       4       64K     system_distributed.service_levels/repair-meta/active
10      0       0B      system_distributed.cdc_streams_descriptions_v2/repair-meta/waiting_for_admission
4       0       0B      system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/waiting_for_admission
1       0       0B      system.group0_history/data-query/waiting_for_admission

25      10      160K    total

Stats:
permit_based_evictions: 0
time_based_evictions: 0
inactive_reads: 0
total_successful_reads: 1189
total_failed_reads: 0
total_reads_shed_due_to_overload: 0
total_reads_killed_due_to_kill_limit: 0
reads_admitted: 2048
reads_enqueued_for_admission: 1357
reads_enqueued_for_memory: 0
reads_admitted_immediately: 706
reads_queued_because_ready_list: 919
reads_queued_because_need_cpu_permits: 10
reads_queued_because_memory_resources: 428
reads_queued_because_count_resources: 0
reads_queued_with_eviction: 0
total_permits: 2063
current_permits: 25
need_cpu_permits: 0
awaits_permits: 0
disk_reads: 0
sstables_read: 0
```

127.0.0.5:
```
INFO  2024-03-04 18:19:33,432 [shard 0:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 163840/1252838604 memory resources: timed out, dumping permit diagnostics:
permits count   memory  table/operation/state
6       6       96K     system_distributed.service_levels/repair-meta/active
4       4       64K     system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/active
1       0       0B      system.group0_history/data-query/waiting_for_admission
10      0       0B      system_distributed.cdc_streams_descriptions_v2/repair-meta/waiting_for_admission
6       0       0B      system_distributed_everywhere.cdc_generation_descriptions_v2/repair-meta/waiting_for_admission

27      10      160K    total

Stats:
permit_based_evictions: 0
time_based_evictions: 0
inactive_reads: 0
total_successful_reads: 1190
total_failed_reads: 0
total_reads_shed_due_to_overload: 0
total_reads_killed_due_to_kill_limit: 0
reads_admitted: 1802
reads_enqueued_for_admission: 1336
reads_enqueued_for_memory: 0
reads_admitted_immediately: 483
reads_queued_because_ready_list: 928
reads_queued_because_need_cpu_permits: 0
reads_queued_because_memory_resources: 408
reads_queued_because_count_resources: 0
reads_queued_with_eviction: 0
total_permits: 1819
current_permits: 27
need_cpu_permits: 0
awaits_permits: 0
disk_reads: 0
sstables_read: 0
```

We need storage people to look at this.

@denesb @bhalevy are these dumps telling anything?

**Comment by kbr-scylla** [comment]
Taking off area/raft label, unlikely to be related. Happens both in Raft mode and during RECOVERY (where Raft is not operating)

**Comment by kbr-scylla** [comment]
[logs.tar.gz](https://github.com/scylladb/scylladb/files/14483744/logs.tar.gz)
Uploaded logs, TRACE enabled in repair module (with some additional logs of mine)

**Comment by kbr-scylla** [comment]
Maybe we're leaking the permits.

**Comment by mykaul** [comment]
```Semaphore streaming with 10/10``` - Is this low on purpose?

**Comment by avikivity** [comment]
> `Semaphore streaming with 10/10` - Is this low on purpose?

Yes, we don't want to waste memory on a streaming task (which doesn't need concurrency to perform well)

**Comment by kbr-scylla** [comment]
> I couldn't reproduce it on my laptop or Warsaw machines.

Okay, actually reproduced on both using the simpler method (5 nodes, shutdown 1, removenode) using your ansible scripts.

Maybe it's something specific to the config you used in the scripts

**Comment by denesb** [comment]
I will have a look. I suspect it has something to do with raft also using this semaphore. I smell a new deadlock.

**Comment by kbr-scylla** [comment]
So it looks like it reproduces more often if we have... more memory?

test.py reproducer: https://github.com/kbr-scylla/scylladb/commit/67cb8f59bfe899c82d8bb1167eb475f86a910580
run with `./test.py --mode dev test_removenode_hang`

It reproduces 9/10 times on one of the strong Warsaw machines.

Note the diff:
```diff
diff --git a/test/pylib/scylla_cluster.py b/test/pylib/scylla_cluster.py
index c7721bd603..1b6bb1af76 100644
--- a/test/pylib/scylla_cluster.py
+++ b/test/pylib/scylla_cluster.py
@@ -121,7 +121,7 @@ def make_scylla_conf(workdir: pathlib.Path, host_addr: str, seed_addrs: List[str
 # Scylla refuses to boot.
 SCYLLA_CMDLINE_OPTIONS = [
     '--smp', '2',
-    '-m', '1G',
+    #'-m', '1G',
     '--collectd', '0',
     '--overprovisioned',
     '--max-networking-io-control-blocks', '1000',
```

i.e. I removed the `-m 1G` from flags.

With it, it does not reproduce.

It also reproduces on my laptop (after removing the memory parameter), but much less often, more like 1/10 times.

**Comment by kbr-scylla** [comment]
/proc/meminfo on my laptop
```
MemTotal:       32703980 kB
MemFree:        15962572 kB
MemAvailable:   20070956 kB
Buffers:           69956 kB
Cached:          6427980 kB
SwapCached:          344 kB
Active:         11277336 kB
Inactive:        3281892 kB
Active(anon):    9602444 kB
Inactive(anon):   656716 kB
Active(file):    1674892 kB
Inactive(file):  2625176 kB
Unevictable:     1180344 kB
Mlocked:            5748 kB
SwapTotal:       8388604 kB
SwapFree:        8233468 kB
Zswap:                 0 kB
Zswapped:              0 kB
Dirty:             18208 kB
Writeback:             0 kB
AnonPages:       9241388 kB
Mapped:          1054068 kB
Shmem:           2192696 kB
KReclaimable:     277256 kB
Slab:             518240 kB
SReclaimable:     277256 kB
SUnreclaim:       240984 kB
KernelStack:       31648 kB
PageTables:        90348 kB
SecPageTables:         0 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    24740592 kB
Committed_AS:   144082092 kB
VmallocTotal:   34359738367 kB
VmallocUsed:      102056 kB
VmallocChunk:          0 kB
Percpu:            17920 kB
HardwareCorrupted:     0 kB
AnonHugePages:     34816 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:         0 kB
FilePmdMapped:         0 kB
CmaTotal:              0 kB
CmaFree:               0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:      242780 kB
DirectMap2M:     7999488 kB
DirectMap1G:    25165824 kB
```


on the Warsaw Office machine
```
MemTotal:       131805848 kB
MemFree:        69299904 kB
MemAvailable:   119045556 kB
Buffers:          106376 kB
Cached:         48753372 kB
SwapCached:            0 kB
Active:         24440224 kB
Inactive:       30413916 kB
Active(anon):    5120000 kB
Inactive(anon):  2799872 kB
Active(file):   19320224 kB
Inactive(file): 27614044 kB
Unevictable:       11656 kB
Mlocked:           11656 kB
SwapTotal:       8388604 kB
SwapFree:        8388604 kB
Zswap:                 0 kB
Zswapped:              0 kB
Dirty:                56 kB
Writeback:             0 kB
AnonPages:       6006152 kB
Mapped:           816252 kB
Shmem:           1920448 kB
KReclaimable:    4058832 kB
Slab:            5550192 kB
SReclaimable:    4058832 kB
SUnreclaim:      1491360 kB
KernelStack:       27248 kB
PageTables:        65452 kB
SecPageTables:         0 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    74291528 kB
Committed_AS:   51195604 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       87400 kB
VmallocChunk:          0 kB
Percpu:            63360 kB
HardwareCorrupted:     0 kB
AnonHugePages:    407552 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:    180224 kB
FilePmdMapped:         0 kB
CmaTotal:              0 kB
CmaFree:               0 kB
Unaccepted:            0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:      302340 kB
DirectMap2M:     5896192 kB
DirectMap1G:    127926272 kB
```

**Comment by kbr-scylla** [comment]
Hypothesis.

Repair coordinator starts concurrent repair tasks, let's say for simplicity there are two tasks, T1 and T2.

T1 needs to take the semaphore on nodes A and B and same for T2. Neither task will progress without taking the semaphore on all nodes.
```cpp
            try {
                parallel_for_each(master.all_nodes(), [&, this] (repair_node_state& ns) {
                    const auto& node = ns.node;
                    ns.state = repair_state::row_level_start_started;
                    return master.repair_row_level_start(node, _shard_task.get_keyspace(), _cf_name, _range, schema_version, _shard_task.reason(), compaction_time, ns.shard).then([&] () {
```
(`repair_row_level_start` RPC takes the semaphore)

T1 manages to take the the semaphore on A.
T2 manages to take the semaphore on B.

And then they get stuck trying to take semaphore on the other node. Deadlock.

(We need more tasks because the semaphore allows concurrency of 10, but you get the idea.)

cc @asias

**Comment by denesb** [comment]
The hyphotesis looks plausible. I think we are hitting this because we have more and more distributed tables and with RBNO decomission, all of those are repaired in parallel.

**Comment by kbr-scylla** [comment]
Reproduces with smp=1 too, might be easier to debug

**Comment by kbr-scylla** [comment]
Added some debug to print which tasks managed to start repair on which followers after a timeout:
```diff
diff --git a/repair/row_level.cc b/repair/row_level.cc
index 19ba702f7e..46b32ca752 100644
--- a/repair/row_level.cc
+++ b/repair/row_level.cc
@@ -3006,10 +3006,16 @@ class row_level_repair {
             std::vector<repair_node_state> nodes_to_stop;
             nodes_to_stop.reserve(master.all_nodes().size());
             try {
+                std::unordered_map<gms::inet_address, bool> started;
+                for (auto& ns: master.all_nodes()) {
+                    started[ns.node] = false;
+                }
+                try{
                 parallel_for_each(master.all_nodes(), [&, this] (repair_node_state& ns) {
                     const auto& node = ns.node;
                     ns.state = repair_state::row_level_start_started;
                     return master.repair_row_level_start(node, _shard_task.get_keyspace(), _cf_name, _range, schema_version, _shard_task.reason(), compaction_time, ns.shard).then([&] () {
+                        started[node] = true;
                         ns.state = repair_state::row_level_start_finished;
                         nodes_to_stop.push_back(ns);
                         ns.state = repair_state::get_estimated_partitions_started;
@@ -3020,6 +3026,10 @@ class row_level_repair {
                         });
                     });
                 }).get();
+                } catch (...) {
+                    rlogger.info("repair range {} seed {} ks {} cf {} failure, started: {}", _range, compaction_time, _shard_task.get_keyspace(), _cf_name, started);
+                    throw;
+                }
 
                 parallel_for_each(master.all_nodes(), [&, this] (repair_node_state& ns) {
                     const auto& node = ns.node;
```



Result:
```
INFO  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair range (5211778979464718027,5475295394935271936] seed   1709643212 ks system_distributed cf view_build_status failure, started: {{127.101.12.6 -> 0}, {127.101.12.13 -> 1}, {127.101.12.1 -> 1}}
WARN  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair[14af4357-f328-4e08-bc8f-8a53a0f7f1f7]: shard=0, keyspace=system_distributed, cf=view_build_status, range=(5211778979464718027,5475295394935271936], got error in row level repair: seastar::rpc::remote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 13:53:42,628 [shard 0:strm] repair - repair_meta::stop
INFO  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair range (5475295394935271936,5821049109740743076] seed   1709643212 ks system_distributed cf view_build_status failure, started: {{127.101.12.6 -> 0}, {127.101.12.13 -> 1}, {127.101.12.1 -> 1}}
WARN  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair[14af4357-f328-4e08-bc8f-8a53a0f7f1f7]: shard=0, keyspace=system_distributed, cf=view_build_status, range=(5475295394935271936,5821049109740743076], got error in row level repair: seastar::rpc::remote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 13:53:42,628 [shard 0:strm] repair - repair_meta::stop
INFO  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair range (-8161013658090232528,-8026740397683004978] seed   1709643212 ks system_distributed cf view_build_status failure, started: {{127.101.12.6 -> 1}, {127.101.12.11 -> 0}, {127.101.12.1 -> 1}}
WARN  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair[14af4357-f328-4e08-bc8f-8a53a0f7f1f7]: shard=0, keyspace=system_distributed, cf=view_build_status, range=(-8161013658090232528,-8026740397683004978], got error in row level repair: seastar::rpc::remote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 13:53:42,628 [shard 0:strm] repair - repair_meta::stop
INFO  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair range (-3548686153359909993,-3170374321624551737] seed   1709643212 ks system_distributed cf view_build_status failure, started: {{127.101.12.6 -> 0}, {127.101.12.11 -> 0}, {127.101.12.1 -> 1}}
WARN  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair[14af4357-f328-4e08-bc8f-8a53a0f7f1f7]: shard=0, keyspace=system_distributed, cf=view_build_status, range=(-3548686153359909993,-3170374321624551737], got error in row level repair: seastar::rpc::remote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 13:53:42,628 [shard 0:strm] repair - repair_meta::stop
INFO  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair range (-1922630904642987722,-1605365022456070787] seed   1709643212 ks system_distributed cf view_build_status failure, started: {{127.101.12.6 -> 0}, {127.101.12.11 -> 0}, {127.101.12.1 -> 1}}
WARN  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair[14af4357-f328-4e08-bc8f-8a53a0f7f1f7]: shard=0, keyspace=system_distributed, cf=view_build_status, range=(-1922630904642987722,-1605365022456070787], got error in row level repair: seastar::rpc::remote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 13:53:42,628 [shard 0:strm] repair - repair_meta::stop
...
INFO  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair range (9175159311641954176,9214047355809997923] seed   1709643212 ks system_distributed cf view_build_status failure, started: {{127.101.12.6 -> 0}, {127.101.12.11 -> 1}, {127.101.12.1 -> 1}}
WARN  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair[14af4357-f328-4e08-bc8f-8a53a0f7f1f7]: shard=0, keyspace=system_distributed, cf=view_build_status, range=(9175159311641954176,9214047355809997923], got error in row level repair: seastar::rpc::remote_verb_error (Semaphore timed out: streaming)
...
INFO  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair range (9089804833794305365,9175159311641954176] seed   1709643212 ks system_distributed cf view_build_status failure, started: {{127.101.12.6 -> 1}, {127.101.12.11 -> 0}, {127.101.12.1 -> 1}}
WARN  2024-03-05 13:53:42,628 [shard 0:strm] repair - repair[14af4357-f328-4e08-bc8f-8a53a0f7f1f7]: shard=0, keyspace=system_distributed, cf=view_build_status, range=(9089804833794305365,9175159311641954176], got error in row level repair: seastar::rpc::remote_verb_error (Semaphore timed out: streaming)
```

**Comment by kbr-scylla** [comment]
Checked if sorting the nodes and starting repairs sequentially fixes the problem
```diff
diff --git a/repair/row_level.cc b/repair/row_level.cc
index 19ba702f7e..39e05cabab 100644
--- a/repair/row_level.cc
+++ b/repair/row_level.cc
@@ -899,6 +899,8 @@ class repair_meta {
             for (unsigned i = 0; i < all_live_peer_nodes.size(); i++) {
                 _all_node_states.push_back(repair_node_state(all_live_peer_nodes[i], all_live_peer_shards[i].value_or(repair_unspecified_shard)));
             }
+            std::sort(_all_node_states.begin(), _all_node_states.end(),
+                    [] (auto& ns1, auto& ns2) { return ns1.node < ns2.node; });
     }
 
     // follower constructor
@@ -3006,20 +3008,30 @@ class row_level_repair {
             std::vector<repair_node_state> nodes_to_stop;
             nodes_to_stop.reserve(master.all_nodes().size());
             try {
-                parallel_for_each(master.all_nodes(), [&, this] (repair_node_state& ns) {
+                std::unordered_map<gms::inet_address, bool> started;
+                for (auto& ns: master.all_nodes()) {
+                    started[ns.node] = false;
+                }
+                try{
+                for (auto& ns: master.all_nodes()) {
                     const auto& node = ns.node;
                     ns.state = repair_state::row_level_start_started;
-                    return master.repair_row_level_start(node, _shard_task.get_keyspace(), _cf_name, _range, schema_version, _shard_task.reason(), compaction_time, ns.shard).then([&] () {
-                        ns.state = repair_state::row_level_start_finished;
-                        nodes_to_stop.push_back(ns);
-                        ns.state = repair_state::get_estimated_partitions_started;
-                        return master.repair_get_estimated_partitions(node, ns.shard).then([this, node, &ns] (uint64_t partitions) {
-                            ns.state = repair_state::get_estimated_partitions_finished;
-                            rlogger.trace("Get repair_get_estimated_partitions for node={}, estimated_partitions={}", node, partitions);
-                            _estimated_partitions += partitions;
-                        });
-                    });
-                }).get();
+                    master.repair_row_level_start(node, _shard_task.get_keyspace(), _cf_name, _range, schema_version, _shard_task.reason(), compaction_time, ns.shard).get();
+
+                    started[node] = true;
+                    ns.state = repair_state::row_level_start_finished;
+                    nodes_to_stop.push_back(ns);
+                    ns.state = repair_state::get_estimated_partitions_started;
+                    uint64_t partitions = master.repair_get_estimated_partitions(node, ns.shard).get();
+
+                    ns.state = repair_state::get_estimated_partitions_finished;
+                    rlogger.trace("Get repair_get_estimated_partitions for node={}, estimated_partitions={}", node, partitions);
+                    _estimated_partitions += partitions;
+                }
+                } catch (...) {
+                    rlogger.info("repair range {} seed {} ks {} cf {} failure, started: {}", _range, compaction_time, _shard_task.get_keyspace(), _cf_name, started);
+                    throw;
+                }
 
                 parallel_for_each(master.all_nodes(), [&, this] (repair_node_state& ns) {
                     const auto& node = ns.node;
```


It doesn't, there are tasks that don't manage to take any lock
and other tasks had to get stuck somewhere else
```
DEBUG 2024-03-05 14:09:03,241 [shard 0:strm] repair - >>> Started Row Level Repair (Follower): local=127.59.134.29, peers=127.59.134.2, repair_meta_id=67, keyspace=system_distributed, cf=view_build_status, schema_version=1e8042e9-b1b2-3dcd-b390-76169d706
660, range=(-3458148616596129131,-3345602840200988012], seed=6846724512834886178, max_row_buf_siz=33554432
DEBUG 2024-03-05 14:09:03,241 [shard 0:strm] repair - >>> Started Row Level Repair (Follower): local=127.59.134.29, peers=127.59.134.2, repair_meta_id=68, keyspace=system_distributed, cf=view_build_status, schema_version=1e8042e9-b1b2-3dcd-b390-76169d706
660, range=(-2896348419043940575,-2730346094994259242], seed=5859411254670184045, max_row_buf_siz=33554432
DEBUG 2024-03-05 14:09:03,241 [shard 0:strm] repair - >>> Started Row Level Repair (Follower): local=127.59.134.29, peers=127.59.134.11, repair_meta_id=46, keyspace=system_distributed, cf=cdc_streams_descriptions_v2, schema_version=bb9f554e-9f62-31d7-87f
3-a1c9ce2cd0a8, range=(-inf, -9104815179056269938], seed=12370814807470936581, max_row_buf_siz=33554432
DEBUG 2024-03-05 14:09:03,241 [shard 0:strm] repair - >>> Started Row Level Repair (Follower): local=127.59.134.29, peers=127.59.134.11, repair_meta_id=47, keyspace=system_distributed, cf=cdc_streams_descriptions_v2, schema_version=bb9f554e-9f62-31d7-87f
3-a1c9ce2cd0a8, range=(-9104815179056269938,-9053314121691844183], seed=7384080857346804643, max_row_buf_siz=33554432
DEBUG 2024-03-05 14:09:03,242 [shard 0:strm] repair - >>> Started Row Level Repair (Follower): local=127.59.134.29, peers=127.59.134.11, repair_meta_id=52, keyspace=system_distributed, cf=cdc_streams_descriptions_v2, schema_version=bb9f554e-9f62-31d7-87f
3-a1c9ce2cd0a8, range=(9222371606588943806,+inf), seed=14729011593865282326, max_row_buf_siz=33554432
DEBUG 2024-03-05 14:09:03,242 [shard 0:strm] repair - >>> Started Row Level Repair (Follower): local=127.59.134.29, peers=127.59.134.2, repair_meta_id=69, keyspace=system_distributed, cf=view_build_status, schema_version=1e8042e9-b1b2-3dcd-b390-76169d706
660, range=(-674671882910943413,-573422689049442166], seed=10005642069867143222, max_row_buf_siz=33554432
INFO  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair range (3407509310361286178,3624660121593504279] seed   1709644143 ks system_distributed cf service_levels failure, started: {{127.59.134.31 -> 0}, {127.59.134.29 -> 0}, {127.59.134.2 -> 0}}
WARN  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair[9c590cbb-f785-4246-ad02-5161ab536336]: shard=0, keyspace=system_distributed, cf=service_levels, range=(3407509310361286178,3624660121593504279], got error in row level repair: seastar::rpc::rem
ote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 14:09:13,239 [shard 0:strm] repair - repair_meta::stop
INFO  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair range (7840221893694606150,7899525292655367939] seed   1709644143 ks system_distributed cf service_levels failure, started: {{127.59.134.31 -> 0}, {127.59.134.29 -> 0}, {127.59.134.2 -> 0}}
WARN  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair[9c590cbb-f785-4246-ad02-5161ab536336]: shard=0, keyspace=system_distributed, cf=service_levels, range=(7840221893694606150,7899525292655367939], got error in row level repair: seastar::rpc::rem
ote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 14:09:13,239 [shard 0:strm] repair - repair_meta::stop
INFO  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair range (-7267431360058142955,-7186864637833239767] seed   1709644143 ks system_distributed cf service_levels failure, started: {{127.59.134.31 -> 0}, {127.59.134.29 -> 0}, {127.59.134.2 -> 0}}
WARN  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair[9c590cbb-f785-4246-ad02-5161ab536336]: shard=0, keyspace=system_distributed, cf=service_levels, range=(-7267431360058142955,-7186864637833239767], got error in row level repair: seastar::rpc::r
emote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 14:09:13,239 [shard 0:strm] repair - repair_meta::stop
INFO  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair range (3318310074404877610,3334164406672711381] seed   1709644143 ks system_distributed cf service_levels failure, started: {{127.59.134.31 -> 0}, {127.59.134.29 -> 0}, {127.59.134.2 -> 0}}
WARN  2024-03-05 14:09:13,239 [shard 0:strm] repair - repair[9c590cbb-f785-4246-ad02-5161ab536336]: shard=0, keyspace=system_distributed, cf=service_levels, range=(3318310074404877610,3334164406672711381], got error in row level repair: seastar::rpc::rem
ote_verb_error (Semaphore timed out: streaming)
DEBUG 2024-03-05 14:09:13,239 [shard 0:strm] repair - repair_meta::stop
```

**Comment by kbr-scylla** [comment]
(or it could be that some tasks manage to finish immediately once the other tasks timeout and unblock the lock)

**Comment by denesb** [comment]
I'm trying to test a simple fix, but the simple fix is getting more and more dirty and it still doesn't work.

**Comment by kbr-scylla** [comment]
@denesb but from what I see, there's more to this than just the deadlock theoretized above.

I see that some tasks manage to get past the semaphore taking stage (i.e. all insert_repair_meta calls succeeded), but then get stuck at some later stage (didn't determine which one yet). Then there are concurrent tasks that get stuck on the semaphore taking stage. Once these tasks timeout (due to my modification to set timeout to 10 seconds), the other tasks (that are stuck on later stage) actually proceed and succeed.

So the question is where are those other tasks getting stuck -- they are holding the permit

**Comment by denesb** [comment]
> So the question is where are those other tasks getting stuck -- they are holding the permit

Once they start reading, their permit becomes evictable. At that point some other repair might evict their permit, so they have to reapply and are queued behind the deadlocked repairs.

**Comment by kbr-scylla** [comment]
But then, with the change I applied (to take locks sequentially in deterministic order), the deadlocked ones should get unstuck as the semaphore becomes available due to the eviction.

But I see them hanging.

After further debugging I determined that some tasks are getting stuck here:
```cpp
            parallel_for_each(nodes_to_stop, [&] (repair_node_state& ns) {
                auto node = ns.node;
                master.set_repair_state(repair_state::row_level_stop_started, node);
                return master.repair_row_level_stop(node, _shard_task.get_keyspace(), _cf_name, _range, ns.shard).then([node, &master] {
                    master.set_repair_state(repair_state::row_level_stop_finished, node);
                });
            }).get();
```

**Comment by kbr-scylla** [comment]
> After further debugging I determined that some tasks are getting stuck here:

Actually no, ignore that

**Comment by denesb** [comment]
@kbr-scylla I have a fix, the test is now passing, can you please also run through your reproducer?
```diff
$ git diff HEAD
diff --git a/repair/row_level.cc b/repair/row_level.cc
index 4011183779..95031b2182 100644
--- a/repair/row_level.cc
+++ b/repair/row_level.cc
@@ -788,6 +788,7 @@ class repair_meta {
     repair_hasher _repair_hasher;
     gc_clock::time_point _compaction_time;
     bool _is_tablet;
+    reader_concurrency_semaphore::inactive_read_handle _fake_ir_handle;
 public:
     std::vector<repair_node_state>& all_nodes() {
         return _all_node_states;
@@ -899,6 +900,7 @@ class repair_meta {
             for (unsigned i = 0; i < all_live_peer_nodes.size(); i++) {
                 _all_node_states.push_back(repair_node_state(all_live_peer_nodes[i], all_live_peer_shards[i].value_or(repair_unspecified_shard)));
             }
+            _fake_ir_handle = _db.local().get_reader_concurrency_semaphore().register_inactive_read(make_empty_flat_reader_v2(_schema, _permit));
     }
 
     // follower constructor
@@ -1137,6 +1139,7 @@ class repair_meta {
         std::list<repair_row> cur_rows;
         std::exception_ptr ex;
         if (!_repair_reader) {
+            _db.local().get_reader_concurrency_semaphore().unregister_inactive_read(std::move(_fake_ir_handle));
             _repair_reader.emplace(_db,
                 _db.local().find_column_family(_schema->id()),
                 _schema,
```

**Comment by denesb** [comment]
With the above fix, I can no longer reproduce the hang with your test. before, I would reproduce every single time.

**Comment by kbr-scylla** [comment]
Ahh there is another place that I completely missed (－‸ლ)

```
            auto permit = _shard_task.db.local().obtain_reader_permit(_shard_task.db.local().find_column_family(_table_id), "repair-meta", db::no_timeout, {}).get();
```

the coordinator/master process itself also takes the same permit

so that's why my fix to apply deterministic order didn't work

It could happen that we start 10 concurrent master tasks which take the permit
and then each of these master does RPC to followers, and master is also one of the followers
and of course the follower side will not be able to take the permit...

This design doesn't make any sense to me

**Comment by kbr-scylla** [comment]
@denesb yeah looks like it works, but should we stop there?

I mean, shouldn't we prevent deadlocks from happening in the first place? Instead of "masking" them by this evicting mechanism. As far as I understand, this is a way to get out of a deadlock, but that deadlock should've never been there IMO.

**Comment by kbr-scylla** [comment]
BTW is this fix backportable?

**Comment by denesb** [comment]
This eviction workaround is the best we have, without a distributed and lineariziable state machine, which manages repair scheduling centrally. We have such a process now -- group0. But at the time we had to tackle this problem, we didn't yet. And so long as we allow users to directly start repairs at will, we have to keep this evictable hack.

Yes, this patch is backportable and I will submit a PR tomorrow.

**Comment by kbr-scylla** [comment]
> I will submit a PR tomorrow

It would be good to have a test.py test which reproduces it more or less consistently even with low memory.

Why is high memory necessary? Because we start more repair tasks concurrently?

Perhaps we can force this in low memory environment too using some error injections, e.g. by artificially increasing concurrency of repair tasks or by lowering the semaphore concurrency

**Comment by kostja** [comment]
* Botond Dénes ***@***.***> [24/03/05 17:44]:

For the uninitiated here and ignorant of "inactive read" concept,
what does the fix actually do?


-- 
Konstantin Osipov, Moscow, Russia

**Comment by denesb** [comment]
> It would be good to have a test.py test which reproduces it more or less consistently even with low memory.
> 
> Why is high memory necessary? Because we start more repair tasks concurrently?

Yes, we have a limit of how much memory a repair can use and  we base our concurrency on this limit, by dividing the target total memory consumption limit by the individual limit. Artificially lowering this limit can maybe help reproduce this. We had no reproducers for any of the previous deadlocks we found, because races are really hard to produce reliably in a controlled environment. When you want a race to happen, it wont, and when you don't want it to happen, it will.

> Perhaps we can force this in low memory environment too using some error injections, e.g. by artificially increasing concurrency of repair tasks or by lowering the semaphore concurrency

Lowering the semaphore concurrency to 1 can also work, maybe.

**Comment by denesb** [comment]
> For the uninitiated here and ignorant of "inactive read" concept, what does the fix actually do?

Inactive read is evictable. The fix effectively marks the permit as evictable as soon as the repair meta object is constructed. Before the fix, the permit would become evictable only after the first read from disk happened.

The interface for marking a permit as inactive is intricately tied to readers, doing this without a reader is very clunky and I will improve that. But for a to-be-backported fix, this patch should be good.

**Comment by denesb** [comment]
> Can we make the permit inactive by default inside the call of obtain_reader_permit()? It is sad we need a fake empty reader to make it inactive.

I will work on a better API to achieve this, as a follow-up. Indeed, this current API is more than clunky.

> We have
> 
> ```
> future<> repair_reader::on_end_of_stream() noexcept {
>     return _reader.close().then([this] {
>         _permit.release_base_resources();
>         _reader = mutation_fragment_v1_stream(make_empty_flat_reader_v2(_schema, _permit));
>         _reader_handle.reset();
>     });         
> } 
> ```
> 
> Do we also need the register_inactive_read trick here for the empty reader?

No, because we call `_permit.release_base_resources();` here, which releases all resource associated with the permit. This is equivalent to marking the permit as evictable and immediately evicting it.

**Comment by asias** [comment]
> > Can we make the permit inactive by default inside the call of obtain_reader_permit()? It is sad we need a fake empty reader to make it inactive.
> 
> I will work on a better API to achieve this, as a follow-up. Indeed, this current API is more than clunky.

Sure.  The current fix is good enough for now and for backport. 

> 
> > We have
> > ```
> > future<> repair_reader::on_end_of_stream() noexcept {
> >     return _reader.close().then([this] {
> >         _permit.release_base_resources();
> >         _reader = mutation_fragment_v1_stream(make_empty_flat_reader_v2(_schema, _permit));
> >         _reader_handle.reset();
> >     });         
> > } 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do we also need the register_inactive_read trick here for the empty reader?
> 
> No, because we call `_permit.release_base_resources();` here, which releases all resource associated with the permit. This is equivalent to marking the permit as evictable and immediately evicting it.

Ok. I think so too, Just to confirming.

**Comment by denesb** [comment]
Fix: https://github.com/scylladb/scylladb/pull/17646

**Comment by denesb** [comment]
I will try to write a reproducer, using regular amount of memory as a separate PR.

**Comment by kostja** [comment]
* Botond Dénes ***@***.***> [24/03/06 10:37]:
> > For the uninitiated here and ignorant of "inactive read" concept, what does the fix actually do?
> Inactive read is evictable. The fix effectively marks the permit as evictable as soon as the repair meta object is constructed. Before the fix, the permit would become evictable only after the first read from disk happened.
> 
> The interface for marking a permit as inactive is intricately tied to readers, doing this without a reader is very clunky and I will improve that. But for a to-be-backported fix, this patch should be good.

But how is it correct? If we make the read "evictable" (what does
it mean btw exactly?) then we effectively remove the limit on
concurrency, so this is as if we didn't have the semaphore in the
first place, no?


-- 
Konstantin Osipov, Moscow, Russia

**Comment by kbr-scylla** [comment]
> Yes, we have a limit of how much memory a repair can use and we base our concurrency on this limit, by dividing the target total memory consumption limit by the individual limit. Artificially lowering this limit can maybe help reproduce this. We had no reproducers for any of the previous deadlocks we found, because races are really hard to produce reliably in a controlled environment. When you want a race to happen, it wont, and when you don't want it to happen, it will.

We don't need a 100% reproducer. The one above was catching the problem in 90% of the runs which is enough. We could probably make it 100% by error-injecting a sleep after the master task takes permit.

**Comment by denesb** [comment]
> * Botond Dénes ***@***.***> [24/03/06 10:37]:
> > For the uninitiated here and ignorant of "inactive read" concept, what does the fix actually do? Inactive read is evictable. The fix effectively marks the permit as evictable as soon as the repair meta object is constructed. Before the fix, the permit would become evictable only after the first read from disk happened. The interface for marking a permit as inactive is intricately tied to readers, doing this without a reader is very clunky and I will improve that. But for a to-be-backported fix, this patch should be good.
> But how is it correct? If we make the read "evictable" (what does it mean btw exactly?) then we effectively remove the limit on concurrency, so this is as if we didn't have the semaphore in the first place, no?

An evicted reader is destroyed. When such a read is resumed, it has to re-acquire the permit, then the reader has to be re-created from the position it was at before being evicted. By eviction, the semaphore enforces the concurrency limit, by destroying readers which are not needed right now.

**Comment by kostja** [comment]
* Botond Dénes ***@***.***> [24/03/06 12:41]:
> > > For the uninitiated here and ignorant of "inactive read" concept, what does the fix actually do? Inactive read is evictable. The fix effectively marks the permit as evictable as soon as the repair meta object is constructed. Before the fix, the permit would become evictable only after the first read from disk happened. The interface for marking a permit as inactive is intricately tied to readers, doing this without a reader is very clunky and I will improve that. But for a to-be-backported fix, this patch should be good.
> > But how is it correct? If we make the read "evictable" (what does it mean btw exactly?) then we effectively remove the limit on concurrency, so this is as if we didn't have the semaphore in the first place, no?
> 
> An evicted reader is destroyed. When such a read is resumed, it
> has to re-acquire the permit, then the reader has to be
> re-created from the position it was at before being evicted. By
> eviction, the semaphore enforces the concurrency limit, by
> destroying readers which are not needed right now.

But doesn't it simply narrow the race window, rather than
eliminate it? I mean, when the reader is resumed, it can preempt,
and another reader will get resumed concurrently, so we still
risk running out of permits anyway? Or the assumption is that if a
reader can not get a permit, it will get preempted and some other
readers will be able to finish, therefore making some more permits
available? This is a nice assumption given there are no other
resources an active reader can block on. If the reader can block
while holding permit on some other semaphore, we can end up in the
same situation.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by denesb** [comment]
> But doesn't it simply narrow the race window, rather than eliminate it? I mean, when the reader is resumed, it can preempt, and another reader will get resumed concurrently, so we still risk running out of permits anyway? Or the assumption is that if a reader can not get a permit, it will get preempted and some other readers will be able to finish, therefore making some more permits available?

Yes.

> This is a nice assumption given there are no other resources an active reader can block on. If the reader can block while holding permit on some other semaphore, we can end up in the same situation.

Yes, if there is another semaphore at play, this will do nothing to prevent deadlocks. The assumption is that this is not the case.

**Comment by kbr-scylla** [comment]
@denesb IIUC this labels-based automation will only work if labels are added before the merge? Dunno (cc @yaronkaikov )
we might need to backport this manually

**Comment by yaronkaikov** [comment]
> > @denesb IIUC this labels-based automation will only work if labels are added before the merge? Dunno (cc @yaronkaikov ) we might need to backport this manually
> 
> It can be added at any time.

But those labels should be added to the PR, not the issue (at least for now)

**Comment by mykaul** [comment]
@yaronkaikov - is the bot supposed to backport this now?

**Comment by yaronkaikov** [comment]
> @yaronkaikov - is the bot supposed to backport this now?

Nope. first we need to merge https://github.com/scylladb/scylladb/pull/17715 which solve it

**Comment by mykaul** [comment]
5.2 backport - https://github.com/scylladb/scylladb/pull/17730

**Comment by denesb** [comment]
5.4 backport - https://github.com/scylladb/scylladb/pull/17950
